python/pylibcudf/pylibcudf/io/types.pyx

```diff
@@ -549,10 +535,31 @@ cdef class SourceInfo:
             raise ValueError("Sources must be a list of str/paths, "
                              "bytes, io.BytesIO, io.StringIO, or a Datasource")
 
-        if empty_buffer is True:
-            c_host_buffers.push_back(host_buffer(<char*>NULL, 0))
+        self.c_obj = source_info(host_span[host_span[const_byte]](self._hspans))
+
+    def _init_byte_like_sources(self, list sources, type expected_type):
+        cdef const unsigned char[::1] c_buffer
+        cdef bint empty_buffer = True
+
+        for src in sources:
+            if not isinstance(src, expected_type):
+                raise ValueError("All sources must be of the same type!")
+
+            if expected_type is bytes:
+                c_buffer = src
+            else:
+                c_buffer = src.getbuffer()
+
+            if c_buffer.shape[0] > 0:
+                self._hspans.push_back(
+                    host_span[const_byte](<const_byte*>&c_buffer[0], c_buffer.shape[0])
+                )
+                empty_buffer = False
 
-        self.c_obj = source_info(c_host_buffers)
+        if empty_buffer:
+            # Not necessary, but future-proofs against libcudf code
```

Comment:
IMO I think it would be better to clarify this with libcudf instead of patching over a possible limitation (and maybe doing something incorrectly)